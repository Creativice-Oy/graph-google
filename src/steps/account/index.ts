import { IntegrationStep } from '@jupiterone/integration-sdk-core';
import { IntegrationConfig, IntegrationStepContext } from '../../types';
import { entities } from '../../constants';
import { createAccountEntity } from './converters';

interface CollectDomainEntityDataResult {
  domainNames: string[];
  primaryDomain?: string;
}

async function collectDomainEntityData(
  context: IntegrationStepContext,
): Promise<CollectDomainEntityDataResult> {
  const domainNames: string[] = [];
  let primaryDomain: string | undefined;

  await context.jobState.iterateEntities(
    {
      _type: entities.DOMAIN._type,
    },
    async (e) => {
      const domainName = e.domainName as string;

      if (e.primary) {
        primaryDomain = domainName;
      }

      domainNames.push(domainName);
      return Promise.resolve();
    },
  );

  return {
    domainNames,
    primaryDomain,
  };
}

export async function createAccount(
  context: IntegrationStepContext,
): Promise<void> {
  const { jobState, instance } = context;
  const { domainNames, primaryDomain } = await collectDomainEntityData(context);

  await jobState.addEntity(
    createAccountEntity({
      account: {
        googleAccountId: instance.config.googleAccountId,
        name: instance.name,
      },
      domainNames,
      primaryDomain,
    }),
  );
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'step-create-account',
    name: 'Account',
    entities: [entities.ACCOUNT],
    relationships: [],
    dependsOn: ['step-fetch-domains'],
    executionHandler: createAccount,
  },
];
