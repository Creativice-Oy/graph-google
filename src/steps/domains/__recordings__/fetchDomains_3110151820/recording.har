{
  "log": {
    "_recordingName": "fetchDomains",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "4.3.0"
    },
    "entries": [
      {
        "_id": "acea721c8193b51ced888cae721cc423",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 977,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "content-length",
              "value": "977"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "www.googleapis.com"
            }
          ],
          "headersSize": 283,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "postData": {
            "mimeType": "application/json",
            "params": [],
            "text": "[REDACTED]"
          },
          "queryString": [],
          "url": "https://www.googleapis.com/oauth2/v4/token"
        },
        "response": {
          "bodySize": 592,
          "content": {
            "mimeType": "application/json; charset=utf-8",
            "size": 592,
            "text": "{\"access_token\":\"[REDACTED]\",\"expires_in\":9999,\"token_type\":\"Bearer\"}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json; charset=utf-8"
            },
            {
              "name": "vary",
              "value": "Origin, X-Origin, Referer"
            },
            {
              "name": "date",
              "value": "Thu, 27 Aug 2020 18:40:46 GMT"
            },
            {
              "name": "server",
              "value": "scaffolding on HTTPServer2"
            },
            {
              "name": "cache-control",
              "value": "private"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "alt-svc",
              "value": "h3-29=\":443\"; ma=2592000,h3-27=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 563,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2020-08-27T18:40:46.678Z",
        "time": 174,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 174
        }
      },
      {
        "_id": "81708a4628a0f4034553d9ea4fc24931",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "x-goog-api-client",
              "value": "gdcl/4.4.0 gl-node/13.10.1 auth/6.0.6"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "google-api-nodejs-client/4.4.0 (gzip)"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "www.googleapis.com"
            }
          ],
          "headersSize": 538,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://www.googleapis.com/admin/directory/v1/customer/C02etu699/domains"
        },
        "response": {
          "bodySize": 975,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 975,
            "text": "{\"kind\":\"admin#directory#domains\",\"etag\":\"\\\"o6ztWZ3lhe-tsV75FsFXpSEZSr0dIx_rEPuaEFoI634/KTGhUSwA2KNfLDZOYUFxZfciSzM\\\"\",\"domains\":[{\"kind\":\"admin#directory#domain\",\"etag\":\"\\\"o6ztWZ3lhe-tsV75FsFXpSEZSr0dIx_rEPuaEFoI634/6j1NZUK3BBPEtHB3FKBAuzFYmEk\\\"\",\"domainName\":\"jupiterone.io\",\"domainAliases\":[{\"kind\":\"admin#directory#domainAlias\",\"etag\":\"\\\"o6ztWZ3lhe-tsV75FsFXpSEZSr0dIx_rEPuaEFoI634/zevudprFOEPK3lTezAE-d43Sy3o\\\"\",\"domainAliasName\":\"jupiterone.io.test-google-a.com\",\"parentDomainName\":\"jupiterone.io\",\"verified\":true,\"creationTime\":\"1549555037874\"}],\"isPrimary\":false,\"verified\":true,\"creationTime\":\"1549555037874\"},{\"kind\":\"admin#directory#domain\",\"etag\":\"\\\"o6ztWZ3lhe-tsV75FsFXpSEZSr0dIx_rEPuaEFoI634/CdS83q1_xnaOGbB-PXkIZoldDVM\\\"\",\"domainName\":\"jupiterone.com\",\"isPrimary\":true,\"verified\":true,\"creationTime\":\"1594817987271\"},{\"kind\":\"admin#directory#domain\",\"etag\":\"\\\"o6ztWZ3lhe-tsV75FsFXpSEZSr0dIx_rEPuaEFoI634/2dHzznB4s2jx9W_p2IM6KWdb0rQ\\\"\",\"domainName\":\"jptr.one\",\"isPrimary\":false,\"verified\":true,\"creationTime\":\"1596544251506\"}]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "etag",
              "value": "\"o6ztWZ3lhe-tsV75FsFXpSEZSr0dIx_rEPuaEFoI634/KTGhUSwA2KNfLDZOYUFxZfciSzM\""
            },
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "vary",
              "value": "Origin, X-Origin, Referer"
            },
            {
              "name": "date",
              "value": "Thu, 27 Aug 2020 18:40:47 GMT"
            },
            {
              "name": "server",
              "value": "ESF"
            },
            {
              "name": "cache-control",
              "value": "private"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "alt-svc",
              "value": "h3-29=\":443\"; ma=2592000,h3-27=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-T050=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 621,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2020-08-27T18:40:46.867Z",
        "time": 242,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 242
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
