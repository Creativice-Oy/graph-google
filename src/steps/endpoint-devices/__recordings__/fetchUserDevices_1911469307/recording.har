{
  "log": {
    "_recordingName": "fetchUserDevices",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "6.0.5"
    },
    "entries": [
      {
        "_id": "acea721c8193b51ced888cae721cc423",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 790,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/x-www-form-urlencoded"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "content-length",
              "value": "790"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "www.googleapis.com"
            }
          ],
          "headersSize": 300,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "postData": {
            "mimeType": "application/x-www-form-urlencoded",
            "params": [],
            "text": "[REDACTED]"
          },
          "queryString": [],
          "url": "https://www.googleapis.com/oauth2/v4/token"
        },
        "response": {
          "bodySize": 455,
          "content": {
            "mimeType": "application/json; charset=utf-8",
            "size": 455,
            "text": "{\"access_token\":\"[REDACTED]\",\"expires_in\":9999,\"token_type\":\"Bearer\"}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "date",
              "value": "Wed, 05 Oct 2022 09:37:34 GMT"
            },
            {
              "name": "expires",
              "value": "Mon, 01 Jan 1990 00:00:00 GMT"
            },
            {
              "name": "cache-control",
              "value": "no-cache, no-store, max-age=0, must-revalidate"
            },
            {
              "name": "pragma",
              "value": "no-cache"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=utf-8"
            },
            {
              "name": "vary",
              "value": "Origin, X-Origin, Referer"
            },
            {
              "name": "server",
              "value": "scaffolding on HTTPServer2"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "alt-svc",
              "value": "h3=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 603,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-10-05T09:37:31.956Z",
        "time": 2724,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 2724
        }
      },
      {
        "_id": "a7befa155e4886a1493596ad23c98206",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "x-goog-api-client",
              "value": "gdcl/5.0.5 gl-node/14.17.6 auth/7.10.3"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "google-api-nodejs-client/5.0.5 (gzip)"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "cloudidentity.googleapis.com"
            }
          ],
          "headersSize": 599,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "view",
              "value": "USER_ASSIGNED_DEVICES"
            },
            {
              "name": "pageSize",
              "value": "200"
            }
          ],
          "url": "https://cloudidentity.googleapis.com/v1/devices?view=USER_ASSIGNED_DEVICES&pageSize=200"
        },
        "response": {
          "bodySize": 748,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 748,
            "text": "{\"devices\":[{\"name\":\"devices/EiQwNjdiODM5MC01MzhiLTQ4ZWYtOWI5NC1kMzM2MmY4NmRkMWU%3D\",\"createTime\":\"2022-09-07T12:50:40.019Z\",\"lastSyncTime\":\"2022-10-04T12:09:24.601Z\",\"ownerType\":\"BYOD\",\"model\":\"Linux\",\"osVersion\":\"Linux\",\"deviceType\":\"LINUX\",\"androidSpecificAttributes\":{},\"deviceId\":\"067b8390-538b-48ef-9b94-d3362f86dd1e\",\"securityPatchTime\":\"1970-01-01T00:00:00Z\"},{\"name\":\"devices/EiQ1Njg5MzFjYi1jYTczLTRiNGEtYTg2MC0yNDgxNTZjMjIyYzE%3D\",\"createTime\":\"2022-09-22T03:01:13Z\",\"lastSyncTime\":\"2022-10-04T04:47:55.399Z\",\"ownerType\":\"BYOD\",\"model\":\"Mac\",\"osVersion\":\"macOS 10.15.7\",\"deviceType\":\"MAC_OS\",\"androidSpecificAttributes\":{},\"deviceId\":\"568931cb-ca73-4b4a-a860-248156c222c1\",\"securityPatchTime\":\"1970-01-01T00:00:00Z\"},{\"name\":\"devices/EiQ2OTE5OTE3NC04OTFjLTQ2YmYtYjNmNi1mNDdiMTlmMmZjZjI%3D\",\"createTime\":\"2022-09-07T13:01:13Z\",\"lastSyncTime\":\"2022-09-07T21:47:53.598Z\",\"ownerType\":\"BYOD\",\"model\":\"Linux\",\"osVersion\":\"Linux\",\"deviceType\":\"LINUX\",\"androidSpecificAttributes\":{},\"deviceId\":\"69199174-891c-46bf-b3f6-f47b19f2fcf2\",\"securityPatchTime\":\"1970-01-01T00:00:00Z\"}],\"nextPageToken\":\"IAE\"}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "vary",
              "value": "Origin, X-Origin, Referer"
            },
            {
              "name": "date",
              "value": "Wed, 05 Oct 2022 09:37:35 GMT"
            },
            {
              "name": "server",
              "value": "ESF"
            },
            {
              "name": "cache-control",
              "value": "private"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "alt-svc",
              "value": "h3=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 483,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-10-05T09:37:34.706Z",
        "time": 463,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 463
        }
      },
      {
        "_id": "08531e33b102cf201d2a608be51127dd",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "x-goog-api-client",
              "value": "gdcl/5.0.5 gl-node/14.17.6 auth/7.10.3"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "google-api-nodejs-client/5.0.5 (gzip)"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "cloudidentity.googleapis.com"
            }
          ],
          "headersSize": 613,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "view",
              "value": "USER_ASSIGNED_DEVICES"
            },
            {
              "name": "pageToken",
              "value": "IAE"
            },
            {
              "name": "pageSize",
              "value": "200"
            }
          ],
          "url": "https://cloudidentity.googleapis.com/v1/devices?view=USER_ASSIGNED_DEVICES&pageToken=IAE&pageSize=200"
        },
        "response": {
          "bodySize": 131,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 131,
            "text": "{}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "vary",
              "value": "Origin, X-Origin, Referer"
            },
            {
              "name": "date",
              "value": "Wed, 05 Oct 2022 09:37:35 GMT"
            },
            {
              "name": "server",
              "value": "ESF"
            },
            {
              "name": "cache-control",
              "value": "private"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "alt-svc",
              "value": "h3=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 483,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-10-05T09:37:35.187Z",
        "time": 414,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 414
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
