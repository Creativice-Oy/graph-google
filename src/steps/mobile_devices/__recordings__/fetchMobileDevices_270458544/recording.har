{
  "log": {
    "_recordingName": "fetchMobileDevices",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "5.1.1"
    },
    "entries": [
      {
        "_id": "acea721c8193b51ced888cae721cc423",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 810,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/x-www-form-urlencoded"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "content-length",
              "value": "810"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "www.googleapis.com"
            }
          ],
          "headersSize": 300,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "postData": {
            "mimeType": "application/x-www-form-urlencoded",
            "params": [],
            "text": "[REDACTED]"
          },
          "queryString": [],
          "url": "https://www.googleapis.com/oauth2/v4/token"
        },
        "response": {
          "bodySize": 549,
          "content": {
            "mimeType": "application/json; charset=utf-8",
            "size": 549,
            "text": "{\"access_token\":\"[REDACTED]\",\"expires_in\":9999,\"token_type\":\"Bearer\"}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "date",
              "value": "Mon, 06 Dec 2021 17:23:49 GMT"
            },
            {
              "name": "cache-control",
              "value": "no-cache, no-store, max-age=0, must-revalidate"
            },
            {
              "name": "expires",
              "value": "Mon, 01 Jan 1990 00:00:00 GMT"
            },
            {
              "name": "pragma",
              "value": "no-cache"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=utf-8"
            },
            {
              "name": "vary",
              "value": "Origin, X-Origin, Referer"
            },
            {
              "name": "server",
              "value": "scaffolding on HTTPServer2"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "alt-svc",
              "value": "h3=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 603,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-12-06T17:23:49.405Z",
        "time": 217,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 217
        }
      },
      {
        "_id": "471330ce9792d6ea1006b47fd7f24633",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "x-goog-api-client",
              "value": "gdcl/4.4.3 gl-node/14.17.3 auth/6.1.6"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "google-api-nodejs-client/4.4.3 (gzip)"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "admin.googleapis.com"
            }
          ],
          "headersSize": 537,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://admin.googleapis.com/admin/directory/v1/customer/C048tgq5f/devices/mobile"
        },
        "response": {
          "bodySize": 331,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 331,
            "text": "{\"kind\":\"admin#directory#mobiledevices\",\"etag\":\"\\\"nqbsbhvoIENh0WbZEZYWTG7mnk2phHz4rrCEo-rHT2k/vyGp6PvFo4RvsFtPoIWeCReyIC8\\\"\"}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "etag",
              "value": "\"nqbsbhvoIENh0WbZEZYWTG7mnk2phHz4rrCEo-rHT2k/vyGp6PvFo4RvsFtPoIWeCReyIC8\""
            },
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "vary",
              "value": "Origin, X-Origin, Referer"
            },
            {
              "name": "date",
              "value": "Mon, 06 Dec 2021 17:23:49 GMT"
            },
            {
              "name": "server",
              "value": "ESF"
            },
            {
              "name": "cache-control",
              "value": "private"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "alt-svc",
              "value": "h3=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 564,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-12-06T17:23:49.637Z",
        "time": 319,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 319
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
