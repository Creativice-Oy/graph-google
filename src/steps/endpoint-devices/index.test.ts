import {
  ExplicitRelationship,
  MappedRelationship,
} from '@jupiterone/integration-sdk-core';
import {
  Recording,
  createMockStepExecutionContext,
  filterGraphObjects,
} from '@jupiterone/integration-sdk-testing';

import { setupIntegrationRecording } from '../../../test/recording';
import { IntegrationConfig } from '../../types';
import { integrationConfig } from '../../../test/config';
import { createAccount } from '../account';
import { fetchUserDevices } from '.';

const tempNewAccountConfig = {
  ...integrationConfig,
  serviceAccountKeyFile: integrationConfig.serviceAccountKeyFile.replace(
    'j1-gc-integration-dev',
    'mineral-liberty-361818',
  ),
  googleAccountId: 'C03yf7qja',
  serviceAccountKeyConfig: {
    ...integrationConfig.serviceAccountKeyConfig,
    project_id: 'mineral-liberty-361818',
  },
};

describe('#fetchUserDevices', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupIntegrationRecording({
      directory: __dirname,
      name: 'fetchUserDevices',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: tempNewAccountConfig,
    });

    await createAccount(context);
    await fetchUserDevices(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    const filteredRelationships = filterGraphObjects(
      context.jobState.collectedRelationships,
      (r) => !r._mapping,
    ) as {
      targets: ExplicitRelationship[];
      rest: MappedRelationship[];
    };

    const mappedRelationships = filteredRelationships.rest;
    expect(mappedRelationships.length).toBeGreaterThan(0);

    expect(
      mappedRelationships
        .filter((e) => e._type === 'google_account_manages_user_endpoint')
        .some(
          (mappedRelationship) =>
            mappedRelationship._key ===
            'google_account_C03yf7qja|manages|FORWARD:_type=user_endpoint:deviceId=067b8390-538b-48ef-9b94-d3362f86dd1e',
        ),
    ).toBe(true);
  });
});
