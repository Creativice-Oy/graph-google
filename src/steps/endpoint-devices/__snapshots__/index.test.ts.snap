// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchUserDevices should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "google_account_C03yf7qja",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "account": Object {
              "googleAccountId": "C03yf7qja",
              "name": "Local Integration",
            },
            "domainNames": Array [],
            "primaryDomain": undefined,
          },
        },
      ],
      "_type": "google_account",
      "accountId": undefined,
      "createdOn": undefined,
      "displayName": "Local Integration",
      "domains": Array [],
      "id": "C03yf7qja",
      "name": "Local Integration",
      "primaryDomain": undefined,
      "vendor": "Google",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "MANAGES",
      "_key": "google_account_C03yf7qja|manages|FORWARD:_type=user_endpoint:deviceId=067b8390-538b-48ef-9b94-d3362f86dd1e",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "skipTargetCreation": false,
        "sourceEntityKey": "google_account_C03yf7qja",
        "targetEntity": Object {
          "_type": "user_endpoint",
          "deviceId": "067b8390-538b-48ef-9b94-d3362f86dd1e",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "deviceId",
          ],
        ],
      },
      "_type": "google_account_manages_user_endpoint",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "MANAGES",
      "_key": "google_account_C03yf7qja|manages|FORWARD:_type=user_endpoint:deviceId=568931cb-ca73-4b4a-a860-248156c222c1",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "skipTargetCreation": false,
        "sourceEntityKey": "google_account_C03yf7qja",
        "targetEntity": Object {
          "_type": "user_endpoint",
          "deviceId": "568931cb-ca73-4b4a-a860-248156c222c1",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "deviceId",
          ],
        ],
      },
      "_type": "google_account_manages_user_endpoint",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "MANAGES",
      "_key": "google_account_C03yf7qja|manages|FORWARD:_type=user_endpoint:deviceId=69199174-891c-46bf-b3f6-f47b19f2fcf2",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "skipTargetCreation": false,
        "sourceEntityKey": "google_account_C03yf7qja",
        "targetEntity": Object {
          "_type": "user_endpoint",
          "deviceId": "69199174-891c-46bf-b3f6-f47b19f2fcf2",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "deviceId",
          ],
        ],
      },
      "_type": "google_account_manages_user_endpoint",
      "displayName": "MANAGES",
    },
  ],
  "encounteredTypes": Array [
    "google_account",
    "google_account_manages_user_endpoint",
  ],
  "numCollectedEntities": 1,
  "numCollectedRelationships": 3,
}
`;
